docs/rfcs/env-vars.md

```diff
@@ -0,0 +1,220 @@
+    '`1`'.
+3.  **Other uses of $** (issue
+    [`contrib#11846`](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/11846)).
+    If a product requires the use of `$` in some field, we would most
+    likely interpret it as an environment variable. This is not
+    intuitive for users.
+
+#### Unexpected type casting
+
+When using the env syntax we parse its value as YAML. Even if you are
+familiar with YAML, because of the implicit type casting rules and the
+way we store intermediate values, we can get unintuitive results.
+
+The most clear example of this is issue
+[*#8565*](https://github.com/open-telemetry/opentelemetry-collector/issues/8565):
+When setting a variable to value `0123` and using it in a string-typed
+field, it will end up as the string `"83"` (where as the user would
+expect the string to be `0123`).
+
+#### We are less restrictive than the Configuration WG
+
+The Configuration WG defines an [*environment variable expansion feature
+for SDK
+configurations*](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/configuration/file-configuration.md#environment-variable-substitution).
+This accepts only non empty alphanumeric + underscore identifiers
+starting with alphabetic or underscore. If the Configuration WG were to
+expand this in the future (e.g. to include other features present in
+Bash-like syntax as in [opentelemetry-specification/pull/3948](https://github.com/open-telemetry/opentelemetry-specification/pull/3948)), we would not be able to expand our braces syntax to
+support new features without breaking users.
+
+## Desired behavior
+
+*This section is written as if the changes were already implemented.*
+
+The Collector supports **two** different syntaxes for environment
+variable resolution:
+
+1.  The *braces syntax*, `${ENV}`.
+2.  The *env provider syntax*, `${env:ENV}`.
+
+These both have **the same character set and behavior**. They both use
+the env provider under the hood.
+
+The naked syntax supported in BASH is not supported in the Collector.
+Escaping is supported by using two dollar signs. Escaping is also
+honored for unsupported identifiers like `${1}`.
+
+### Type casting rules
+
+The environment variable value is parsed by the yaml.v3 parser to an
+any-typed variable and the original representation as a string is stored
+for numeric types. The `yaml.v3` parser mostly follows the YAML v1.2
+specification with [*some
+exceptions*](https://github.com/go-yaml/yaml#compatibility). You can see
+how it works for some edge cases in this example:
+[*https://go.dev/play/p/RtPmH8aZA1X*](https://go.dev/play/p/RtPmH8aZA1X).
+
+When unmarshalling, we use mapstructure with WeaklyTypedInput
+**disabled**. We check via a hook an `AsString` method from confmap.Conf
+and use its return value when it is valid and we are mapping to a string
+field. This method has default casting rules for unambiguous scalar
```

Comment:
Curious about the motivation here. My preference would be not doing any casting but simply returning the original representation for fields expecting string type, e.g. `0123` -> `"0123"` rather than `"83"`

Reply:
This is what I am proposing (see the table at the bottom of the document). Would you word it differently to make it clearer?

Reply:
Oops I mis-read the 2nd row in the examples. I think this is fine then, maybe mention the examples below on how this works.

Reply:
I explicitly mention the comparison table in the body of the text now :)